### PR TITLE
Added asyncdr test with PX upgrade

### PR DESCRIPTION
**Pull request by gejain**
Added asyncdr test with PX upgrade

**What this PR does / why we need it**:
Added asyncdr test with PX upgrade

**Which issue(s) this PR fixes** (optional)
Closes #https://purestorage.atlassian.net/browse/PTX-24962

**Special notes for your reviewer**:
Green run using job -
https://jenkins.pwx.dev.purestorage.com/job/Stork/view/Stork%20System%20Test/job/StorkAsyncDrWithPXUpgrade/152/

### CODE REVIEW

File: tests/basic/async_dr_test.go [hunk 1]
```diff
@@ -637,13 +639,274 @@ var _ = Describe("{StorkctlPerformFailoverFailbackeckEsClusterwide}", func() {
...
+			kubeConfigWritten = true
+		}
+		wantAllAfterSuiteActions = false
+	})
+
+	JustBeforeEach(func() {
+		upgradeHopsList := make(map[string]string)
+		upgradeHopsList["upgradeHops"] = Inst().UpgradeStorageDriverEndpointList
+		upgradeHopsList["upgradeVolumeDriver"] = "true"
+		StartTorpedoTest("UpgradeVolumeDriver", "Validating volume driver upgrade", upgradeHopsList, 0)
```
Comment (kphalgun-px) [inline tests/basic/async_dr_test.go:663]: UpgradeVolumeDriverDuringMigration?

File: tests/basic/async_dr_test.go [hunk 2]
```diff
@@ -637,13 +639,274 @@ var _ = Describe("{StorkctlPerformFailoverFailbackeckEsClusterwide}", func() {
...
+		log.FailOnError(err, "Refresh Driver end points failed")
+
+		Step("Destroy apps", func() {
+			log.InfoD("Destroy apps")
+			opts := make(map[string]bool)
+			opts[scheduler.OptionsWaitForResourceLeakCleanup] = true
+			for _, ctx := range contexts {
+				TearDownContext(ctx, opts)
+			}
+		})
```
Comment (kphalgun-px) [inline tests/basic/async_dr_test.go:871]: Should we stop all migrations and then destroy apps?

File: tests/basic/async_dr_test.go [hunk 3]
```diff
@@ -637,13 +639,274 @@ var _ = Describe("{StorkctlPerformFailoverFailbackeckEsClusterwide}", func() {
...
+				err = Inst().V.RefreshDriverEndpoints()
+				log.FailOnError(err, "Refresh Driver end points failed")
+				majorVersion := strings.Split(currPXVersion, "-")[0]
+				statsData := make(map[string]string)
+				statsData["numOfNodes"] = fmt.Sprintf("%d", numOfNodes)
+				statsData["fromVersion"] = currPXVersion
+				statsData["toVersion"] = updatedPXVersion
+				statsData["duration"] = fmt.Sprintf("%d mins", durationInMins)
+				statsData["status"] = upgradeStatus
+				dash.UpdateStats("px-upgrade-stats", "px-enterprise", "upgrade", majorVersion, statsData)
```
Comment (kphalgun-px) [inline tests/basic/async_dr_test.go:806]: Can we have something like?

```go
clusters := []struct {
    kubeConfigPath string
    clusterName    string
    storageNodes   []string
}{
    {kubeConfigPathDest, asyncdr.SecondCluster, storageNodesDest},
    {kubeConfigPathSrc, asyncdr.FirstCluster, storageNodes},
}

for _, cluster := range clusters {
    err := hardSetConfig(cluster.kubeConfigPath)
    log.FailOnError(err, fmt.Sprintf("Switching context to cluster with kubeConfigPath: %s failed", cluster.kubeConfigPath))
    err = SetCustomKubeConfig(cluster.clusterName)
    log.FailOnError(err, fmt.Sprintf("Switching context to cluster with kubeConfigPath: %s failed", cluster.kubeConfigPath))
    err = Inst().S.RefreshNodeRegistry()
    log.FailOnError(err, fmt.Sprintf("Node registry refresh failed for cluster with kubeConfigPath: %s", cluster.kubeConfigPath))
    err = Inst().V.RefreshDriverEndpoints()
    log.FailOnError(err, fmt.Sprintf("Refresh Driver end points failed for cluster with kubeConfigPath: %s", cluster.kubeConfigPath))

    currPXVersion, err := Inst().V.GetDriverVersionOnNode(cluster.storageNodes[0])
    if err != nil {
        log.Warnf("error getting driver version for cluster with kubeConfigPath: %s, Err: %v", cluster.kubeConfigPath, err)
    }

    upgradeStatus, updatedPXVersion, durationInMins := upgradePX(upgradeHop, cluster.storageNodes)

    log.InfoD("current PX version: %s, no of nodes: %d, upgrade status: %s, updated px version: %s, duration in mins: %d on cluster with kubeConfigPath: %s",
        currPXVersion, len(cluster.storageNodes), upgradeStatus, updatedPXVersion, durationInMins, cluster.kubeConfigPath)

    majorVersion := strings.Split(currPXVersion, "-")[0]
    statsData := make(map[string]string)
    statsData["numOfNodes"] = fmt.Sprintf("%d", len(cluster.storageNodes))
    statsData["fromVersion"] = currPXVersion
    statsData["toVersion"] = updatedPXVersion
    statsData["duration"] = fmt.Sprintf("%d mins", durationInMins)
    statsData["status"] = upgradeStatus

    dash.UpdateStats("px-upgrade-stats", "px-enterprise", "upgrade", majorVersion, statsData)
}
```

File: tests/basic/async_dr_test.go [hunk 4]
```diff
@@ -637,13 +639,274 @@ var _ = Describe("{StorkctlPerformFailoverFailbackeckEsClusterwide}", func() {
...
 	})
-	
+
 	JustAfterEach(func() {
 		defer EndTorpedoTest()
 		AfterEachTest(contexts, testrailID, runID)
 	})
 })
 
+var _ = Describe("{UpgradeVolumeDriverDuringMigration}", func() {
```
Comment (kphalgun-px) [inline tests/basic/async_dr_test.go:649]: Can we use for-loops wherever possible for the First and Second clusters to reduce redundancy?

File: tests/basic/async_dr_test.go [hunk 5]
```diff
@@ -637,13 +639,274 @@ var _ = Describe("{StorkctlPerformFailoverFailbackeckEsClusterwide}", func() {
...
+		err = hardSetConfig(kubeConfigPathDest)
+		log.FailOnError(err, "Switching context to destination cluster failed")
+		err = SetCustomKubeConfig(asyncdr.SecondCluster)
+		log.FailOnError(err, "Switching context to destination cluster failed")
+
+		err = Inst().S.RefreshNodeRegistry()
+		log.FailOnError(err, "Node registry refresh failed")
+		err = Inst().V.RefreshDriverEndpoints()
+		log.FailOnError(err, "Refresh Driver end points failed")
+		storageNodesDest := node.GetStorageNodes()
```
Comment (kphalgun-px) [inline tests/basic/async_dr_test.go:731]: Can we have map with key as cluster name and value as node array?

File: tests/basic/async_dr_test.go (hunk 2 above)
Comment (gejain) [reply tests/basic/async_dr_test.go:871]: Migrations suspended both side automatically and destination apps already down. So I think cleaning up just source context is enough here. I will discuss once and show you. Thanks.

File: tests/basic/async_dr_test.go (hunk 3 above)
Comment (gejain) [reply tests/basic/async_dr_test.go]: Thank you, just did changes with your help.

File: tests/basic/async_dr_test.go (hunk 4 above)
Comment (gejain) [reply tests/basic/async_dr_test.go]: Done!

File: tests/basic/async_dr_test.go (hunk 5 above)
Comment (gejain) [reply tests/basic/async_dr_test.go]: Done!

File: tests/basic/async_dr_test.go (hunk 1 above)
Comment (gejain) [reply tests/basic/async_dr_test.go]: Done!

**Comment by gejain** [comment]
New Run is Successful with new change @kphalgun-px - https://jenkins.pwx.dev.purestorage.com/job/Stork/view/Stork%20System%20Test/job/StorkAsyncDrWithPXUpgrade/

**Review by Rohit-PX — APPROVED** [review]
@gejain please add a test case in testrail for this.

**Comment by gejain** [comment]
@kphalgun-px - New run with 2 torpedo instances - https://jenkins.pwx.dev.purestorage.com/job/Stork/view/Stork%20System%20Test/job/StorkAsyncDrWithPXUpgrade/156/

Please approve PR.

**Review by kphalgun-px — APPROVED** [review]
LGTM

**Pull request merged**